### PR TITLE
Add rule patching function

### PR DESCRIPTION
**Pull request by raptorsun**
## Description

This pull request adds rule patching function allowing exclusion and patching of Prometheus rules, resolving [issue#697](https://github.com/prometheus-operator/kube-prometheus/issues/697).

For example, the code below will delete the rule "PrometheusOperatorListErrors", modify the rule "PrometheusOperatorWatchErrors" with severity "info" and delete all rule from rule group "group-to-remove".

```
local kp =
  (import 'kube-prometheus/main.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
  };


local rulePatches = {
  excludedRuleGroups: [
    'group-to-remove',
  ],

  excludedRules: [
    {
      name: 'prometheus-operator',
      rules: [
        { alert: 'PrometheusOperatorListErrors' },
      ],
    },
  ],
  patchedRules: [
    {
      name: 'prometheus-operator',
      rules: [
        {
          alert: 'PrometheusOperatorWatchErrors',
          labels: {
            severity: 'info',
          },
        },
      ],
    },
  ],
};

local sanitizeAlertRules = (import 'kube-prometheus/lib/rule-patcher.libsonnet')(rulePatches).sanitizeAlertRules;


sanitizeAlertRules({ 'setup/0namespace-namespace': kp.kubePrometheus.namespace } +
                   {
                     ['setup/prometheus-operator-' + name]: kp.prometheusOperator[name]
                     for name in std.filter((function(name) name != 'serviceMonitor' && name != 'prometheusRule'), std.objectFields(kp.prometheusOperator))
                   } +
                   // serviceMonitor and prometheusRule are separated so that they can be created after the CRDs are ready
                   { 'prometheus-operator-serviceMonitor': kp.prometheusOperator.serviceMonitor } +
                   { 'prometheus-operator-prometheusRule': kp.prometheusOperator.prometheusRule } +
                   { 'kube-prometheus-prometheusRule': kp.kubePrometheus.prometheusRule } +
                   { ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
                   { ['blackbox-exporter-' + name]: kp.blackboxExporter[name] for name in std.objectFields(kp.blackboxExporter) } +
                   { ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) } +
                   { ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
                   { ['kubernetes-' + name]: kp.kubernetesControlPlane[name] for name in std.objectFields(kp.kubernetesControlPlane) }
                   { ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
                   { ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
                   { ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) })
```

The function patchOrExcludeRule uses an additional field "index" to distinguish different rules under the same name (alert/record field is considered as its name). Rules under the same group are ordered by alphabetic order in output.

For example, the following patch will add the field "patch:'A1' to the second rule of "alert:'A'".

```
local patchedRules = [
  {
    name: 'alertmanager.rules',
    rules: [
      {
        alert: 'A',
        index: 1,
        patch: 'A1',
        labels: {
          severity: 'warning',
        },
      },
      {
        alert: 'B',
        patch: 'B0',
        labels: {
          severity: 'warning',
        },
      },
      {
        alert: 'C',
        patch: 'C0',
        labels: {
          severity: 'warning',
        },
      },
      {
        alert: 'D',
        patch: 'D0',
        labels: {
          severity: 'warning',
        },
      },
    ],
  },
];

local rulesToPatch = [
  {
    name: 'alertmanager.rules',
    rules: [
      {
        alert: 'A',
        field: 'A0 rule',
        labels: {
          severity: 'warning',
        },
      },
      {
        alert: 'A',
        field: 'A1 rule',
        labels: {
          severity: 'warning',
        },
      },
      {
        alert: 'B',
        field: 'B0 rule',
        labels: {
          severity: 'warning',
        },
      },
    ],
  },
];
```

## Type of change

_What type of changes does your code introduce to the kube-prometheus? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. Later this will be copied to the changelog file._

```release-note
Add rule patching function allowing exclusion and patching of Prometheus rules.
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -229,23 +229,55 @@ local kp =
     },
   };
 
-{ 'setup/0namespace-namespace': kp.kubePrometheus.namespace } +
```
Comment (ArthurSens) [inline README.md:232]: There is no need to change the README to add those functions. I think README needs to include only the very basics for someone to get started with kube-prometheus, and modifying alerts is not something that everyone will need on their first day.

Also, this snippet is auto-generated when running `make generate`, that's why CI is failing 🙂

File: build.sh [hunk 2]
```diff
@@ -7,6 +7,11 @@ set -x
 # only exit with zero if all commands of the pipeline exit successfully
 set -o pipefail
 
+if [ ! -f "$1" ]; then
+    echo "$1 is not a file"
+    exit 1
+fi
+
```
Comment (ArthurSens) [inline build.sh:14]: Why is this necessary?

File: docs/developing-prometheus-rules-and-grafana-dashboards.md [hunk 3]
```diff
@@ -39,23 +39,55 @@ local kp =
     },
   };
 
-{ 'setup/0namespace-namespace': kp.kubePrometheus.namespace } +
```
Comment (ArthurSens) [inline docs/developing-prometheus-rules-and-grafana-dashboards.md:42]: Likewise the comment on README, this change doesn't seem to be right. 

It makes perfect sense to document the functions you're adding though, maybe you could update the [Changing Default Rules section](https://github.com/prometheus-operator/kube-prometheus/blob/main/docs/developing-prometheus-rules-and-grafana-dashboards.md#changing-default-rules)?

**Review by ArthurSens — COMMENTED** [review]
Thanks for the PR! A helper function to modify alerts are very much appreciated 🙂

It seems that this PR is still WIP since it has some compiling issues, so I'll just leave one comment unrelated to the jsonnet code

File: README.md (hunk 1 above)
Comment (raptorsun) [reply README.md:232]: Thanks for the review :) I will restore this section and move these descriptions to "Changing Default Rules" section.

File: build.sh (hunk 2 above)
Comment (raptorsun) [reply build.sh:14]: This is not necessary at all for this feature. 
But it would be great to avoid stuffing the directory with many empty files if it does not receive a correct file as the first argument.

File: docs/developing-prometheus-rules-and-grafana-dashboards.md (hunk 3 above)
Comment (raptorsun) [reply docs/developing-prometheus-rules-and-grafana-dashboards.md:42]: This will be moved to the section "Changing Default Rules".

**Comment by raptorsun** [comment]
Hello @ArthurSens , the description of the rule patching function has been moved to the section Changing Default Rules with a code snippet showing how to use it. 
If the argument check for build.sh should be removed, just leave a message and I will tick that commit out.

File: build.sh (hunk 2 above)
Comment (dgrisonnet) [reply build.sh:14]: This change defeats the purpose of defaulting to `example.jsonnet`. I am curious how did you end up with empty files cause it shouldn't happen. That said even if we wanted to add this improvement, I would prefer if we did so in a separate PR since it is unrelated.

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    local sortedRules = std.sort(
+      group.rules, keyF=function(rule)
+        if ('alert' in rule) then
+          rule.alert
+        else if ('record' in rule) then
+          rule.record
+        else
+          assert false : 'rule should have either "alert" or "record" field' + std.toString(rule);
+          ''
+    );
```
Comment (dgrisonnet) [inline jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:113]: can we move that to its own function?

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  else
+    if (lastRule == null) then
+      local updatedRule = std.mergePatch(ruleSet[0], { index: 0 });
+      [updatedRule] + indexRules(updatedRule, ruleSet[1:])
+    else if sameRuleName(lastRule, ruleSet[0]) then
+      local updatedRule = std.mergePatch(ruleSet[0], { index: lastRule.index + 1 });
+      [updatedRule] + indexRules(updatedRule, ruleSet[1:])
+    else
+      local updatedRule = std.mergePatch(ruleSet[0], { index: 0 });
+      [updatedRule] + indexRules(updatedRule, ruleSet[1:]);
```
Comment (dgrisonnet) [inline jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:97]: The `if` and `else` body are identical, we should be able to refactor them.

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    [] + patchOrExcludeRule(rule, ruleSet[1:], operation);
+
+
+local sameRuleName(rule1, rule2) =
+  if ('alert' in rule1 && 'alert' in rule2) && (rule1.alert == rule2.alert) then
+    true
+  else if ('record' in rule1 && 'record' in rule2) && (rule1.record == rule2.record) then
+    true
+  else
+    false;
```
Comment (dgrisonnet) [inline jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:83]: We can dirrectly return the comparaison:
```suggestion
  if ('alert' in rule1 && 'alert' in rule2) then
    rule1.alert == rule2.alert
  else if ('record' in rule1 && 'record' in rule2) then
    rule1.record == rule2.record
  else
    false;
```

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  else if ((('alert' in rule && 'alert' in ruleSet[0]) && std.startsWith(rule.alert, ruleSet[0].alert)) ||
+           (('record' in rule && 'record' in ruleSet[0]) && std.startsWith(rule.record, ruleSet[0].record))) &&
+          (!('index' in ruleSet[0]) || (('index' in ruleSet[0]) && (ruleSet[0].index == rule.index))) then
+    if operation == 'patch' then
+      local patch = {
+        [k]: ruleSet[0][k]
+        for k in std.objectFields(ruleSet[0])
+        if k != 'alert' && k != 'record' && k != 'index'
+      };
+      std.trace('rule match:' + std.toString(ruleSet[0]) + std.toString(rule), [deleteIndex(std.mergePatch(rule, patch))])
```
[the remark this answers is not in the document]
Comment (raptorsun) [reply jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:69]: oops. that slipped my eyes, it will be removed 😶

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  for k in std.objectFields(rule)
+  if k != 'index'
+};
+
+
+local patchOrExcludeRule(rule, ruleSet, operation) =
+  if std.length(ruleSet) == 0 then
+    [deleteIndex(rule)]
+  else if ((('alert' in rule && 'alert' in ruleSet[0]) && std.startsWith(rule.alert, ruleSet[0].alert)) ||
+           (('record' in rule && 'record' in ruleSet[0]) && std.startsWith(rule.record, ruleSet[0].record))) &&
```
[the remark this answers is not in the document]
Comment (raptorsun) [reply jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:61]: I will put a comment in this function and in the markdown file explaining how rules are matched.

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    );
+    local indexedRules = indexRules(null, sortedRules);
+    [patchOrExcludeRule(rule, groupSet[0].rules, operation) for rule in indexedRules]
+  else
+    [] + patchOrExcludeRuleGroup(group, groupSet[1:], operation);
+
+function(params) {
+  local am = self,
+  _config:: defaults + params,
+  assert std.isArray(am._config.excludedRuleGroups) || std.isArray(am._config.excludedRules) || std.isArray(am._config.patchedRules) : 'rule-patcher: excludedRuleGroups, excludedRules or patchedRules should exist in parameter as arrays',
```
[the remark this answers is not in the document]
Comment (raptorsun) [reply jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:122]: it will be divided into 3 assert statements

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet (hunk 4 above)
Comment (raptorsun) [reply jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:113]: yes, it will have a function to extract rule name.

File: jsonnet/kube-prometheus/lib/rule-patcher.libsonnet (hunk 5 above)
Comment (raptorsun) [reply jsonnet/kube-prometheus/lib/rule-patcher.libsonnet:97]: 1st case and 3rd case will be merged together

**Comment by simonpasquier** [comment]
I guess that you can bump `github.com/google/go-jsonnet` since we're already pinned to a post v0.17.0 commit.

https://github.com/prometheus-operator/kube-prometheus/blob/167caa3bd73b9edcd3a68d157cb4f0d535611e2f/scripts/go.mod#L8

**Comment by raptorsun** [comment]
Code is just updated: 

- library is named `rule-sanitizer.libsonnet`
- only the function `sanitizeAlertRules `is exported.
- rule patching example has its rule patches in a dedicated file.

**Comment by raptorsun** [comment]
Linter problem will be fixed by #1425

**Comment by raptorsun** [comment]
The merge conflict has been resolved, ready to review again :) 
@dgrisonnet

**Review by dgrisonnet — APPROVED** [review]
This new version looks good to me.

@paulfantom could you please have another look at this PR?

**Review by PhilipGough — APPROVED** [review]
This lgtm but will give @paulfantom a chance to review

**Review by paulfantom — APPROVED** [review]
lgtm :shipit:
